### PR TITLE
[core][compiled graphs] Adjust overlap algorithm

### LINKED ISSUE #48953: [core][compiled graphs] Overlap Algorithm Adjustment
### Description

Our current overlap algorithm works by moving a candidate READ (NCCL) operation before a COMPUTE operation. If there are other READ (NCCL) or WRITE (NCCL) operations in between, the swap is skipped to avoid deadlock.

We can make the algorithm less conservative by removing the constraint for a WRITE (NCCL) operation in between. Specifically, if there is a WRITE (NCCL) operation in between, and if its peer rank is different from the peer rank of the candidate READ (NCCL) operation, we still allow the swap to happen.

### Use case

_No response_

### PR DESCRIPTION
**Pull request by ruisearch42**
## Why are these changes needed?

Adjust gpu communication overlap algorithm to be less conservative, see details in #48953

## Related issue number

Closes #48953
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/dag/dag_node_operation.py [hunk 1]
```diff
@@ -215,6 +227,25 @@ def is_nccl_write(self) -> bool:
...
+        """
+        Check if this is a NCCL read operation.
+        """
+        return self.operation.type == _DAGNodeOperationType.READ and self.requires_nccl
+
+    @property
+    def is_nccl_write_to_upstream_actor(self) -> bool:
+        """
+        Check if this is a NCCL write operation that writes to an actor
+        that is also one of its immediate upstream NCCL actors.
```
Comment (stephanie-wang) [inline python/ray/dag/dag_node_operation.py:241]: I found this logic hard to understand. Reading the issue, would it work if we instead modified this to `is_nccl_write_to_peer(self, upstream_actor)`, and make sure the caller passes in the right actor?

File: python/ray/dag/compiled_dag_node.py [hunk 2]
```diff
@@ -1682,11 +1682,17 @@ def _generate_dag_operation_graph_node(
                 method_name = exec_task.method_name
                 actor_handle = dag_node._get_actor_handle()
                 requires_nccl = dag_node.type_hint.requires_nccl()
+                downstream_nccl_actors = (
```
Comment (stephanie-wang) [inline python/ray/dag/compiled_dag_node.py:1685]: Why does `downstream_nccl_actors` check whether the current node requires NCCL while `uspstream_nccl_actors` checks whether the upstream node requires NCCL?

**Review by stephanie-wang — COMMENTED** [review]
Sorry, I didn't really understand the upstream/downstream NCCL actors change. Can you either clarify with some examples/comments or see if the suggested change could work instead?

File: python/ray/dag/dag_node_operation.py (hunk 1 above)
Comment (ruisearch42) [reply python/ray/dag/dag_node_operation.py:241]: (There may be more than one upstream_actors, so we need to pass them all in.)
`is_nccl_write_to_upstream_actor` is a short hand and makes the caller code more concise. We can change it to the following at caller side (dag_node_operation.py:792) if you find it easier to understand:

Original:
`overlapped_schedule[j].is_nccl_write_to_upstream_actor`
New:
`overlapped_schedule[j].is_nccl_write_to_peer(overlapped_schedule[j].upstream_actors)`

**Comment by ruisearch42** [comment]
> Sorry, I didn't really understand the upstream/downstream NCCL actors change. Can you either clarify with some examples/comments or see if the suggested change could work instead?

Sure, for the `test_torch_tensor_dag.py::test_torch_tensor_nccl_overlap2` example, we have the following overlapped schedule:
compiled_graph_schedulex

For the Actor_520f.send method, its downstream_nccl_actors: {Actor_e171}, its upstream_nccl_actors: {} (No actor is writing to this node using NCCL). Therefore we can overlap (see the red ovals).

**Comment by ruisearch42** [comment]
The relax algorithm would deadlock for https://github.com/ray-project/ray/pull/49005/commits/5d35b71b536610249516b38d56bc8466d4e494c1
See https://github.com/ray-project/ray/issues/48953#issuecomment-2565764147

**Pull request closed without merging**